django/contrib/admin/options.py

```diff
@@ -309,6 +309,19 @@ def get_readonly_fields(self, request, obj=None):
         """
         return self.readonly_fields
 
+    def get_uneditable_fields(self, request, obj=None):
+        """
+        Get fields that are either readonly or hasn't change/add permissions
+        """
+        if hasattr(request, 'user') and (
+                self.has_change_permission(request, obj) or (self.has_add_permission(request) and not obj)):
+            return self.get_readonly_fields(request, obj)
+        elif hasattr(self, 'declared_fieldsets') and self.declared_fieldsets:
+            return flatten_fieldsets(self.declared_fieldsets)
+        else:
+            return [field.name for field in self.opts.local_fields] + \
+                [field.name for field in self.opts.local_many_to_many]
```

Comment:
This code ignores the value of `readonly_fields`. I noticed this because it breaks an inline with a field which is defined as a method in the inline itself (and, thus, needs to be in `readonly_fields`). I'm currently using this workaround (there might be a more clever way to solve it):


```python
        return [field.name for field in self.opts.local_fields] + \
               [field.name for field in self.opts.local_many_to_many] + \
               list(self.get_readonly_fields(request, obj))
```